pkg/config/configfile.go

```diff
@@ -0,0 +1,369 @@
+package config
+
+import (
+	"errors"
+	"os"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/kubecost/cost-model/pkg/log"
+	"github.com/kubecost/cost-model/pkg/storage"
+)
+
+// HandlerID is a unique identifier assigned to a provided ConfigChangedHandler. This is used to remove a handler
+// from the ConfigFile when it is no longer needed.
+type HandlerID string
+
+//--------------------------------------------------------------------------
+//  ChangeType
+//--------------------------------------------------------------------------
+
+// ChangeType is used to specifically categorize the change that was made on a ConfigFile
+type ChangeType string
+
+// ChangeType constants contain the different types of updates passed through the ConfigChangedHandler
+const (
+	ChangeTypeCreated  ChangeType = "created"
+	ChangeTypeModified ChangeType = "modified"
+	ChangeTypeDeleted  ChangeType = "deleted"
+)
+
+//--------------------------------------------------------------------------
+//  ConfigChangedHandler
+//--------------------------------------------------------------------------
+
+// ConfigChangedHandler is the func handler used to receive change updates about the
+// config file. Both ChangeTypeCreated and ChangeTypeModified yield a valid []byte, while
+// ChangeTypeDeleted yields a nil []byte.
+type ConfigChangedHandler func(ChangeType, []byte)
+
+//--------------------------------------------------------------------------
+//  ConfigFile
+//--------------------------------------------------------------------------
+
+// DefaultHandlerPriority is used as the priority for any handlers added via AddChangeHandler
+const DefaultHandlerPriority int = 1000
+
+// NoBackingStore error is used when the config file's backing storage is missing
+var NoBackingStore error = errors.New("Backing storage does not exist.")
+
+// ConfigFile is representation of a configuration file that can be written to, read, and watched
+// for updates
+type ConfigFile struct {
+	store      storage.Storage
+	file       string
+	dataLock   *sync.Mutex
+	data       []byte
```

Comment:
`Storage` is an abstraction to retrieve data. `data` is the file data. It _is_ cached currently by default. I would like to make this configurable, but it's more important to protect against constant Reads() from S3.

Reply:
> Storage is an abstraction to retrieve data. data is the file data

Data is a possibly-expired cache of the data, meaning it has no consistency guarantee, yeah? I was just hoping we could make that clear, either on the `Read()` method, here on the `data` field, or on the struct itself.

> I would like to make this configurable

Curious what you mean by this, could you explain?

Reply:
There are lots of possibilities, many of them require very obscure use-cases. I can't make any guarantees, nor is that important to the current implementation of this data type. 

Do I think there exists some use case where constantly hitting the file system or s3 might be necessary? Yes, which is why there is a Refresh(), and why there is an Exists(). Most of the current application of this object are leveraging the cache and update handling. It might be nice in the future to allow a configurable way of managing the internal handling of the cache (a policy or something similar), but it's not applicable for the function of the PR as a whole.